evaluator/src/test/kotlin/EvaluatorTest.kt

```diff
@@ -116,4 +117,51 @@ class EvaluatorTest : WordSpec({
             }
         }
     }
+
+    "OSADL compliance rules" should {
+        "return no violation for compatible licenses" {
+            val compatibleOrtResult = ortResult {
+                project("Maven:group:project-name:1") {
+                    license = "AGPL-3.0-only"
+
+                    pkg("Maven:group:package-name:1") {
+                        license = "AGPL-3.0-only"
+
+                        pkg("Maven:group:package-name:2") {
+                            license = "AGPL-3.0-or-later"
+                        }
+                    }
+                }
+            }
+            val script = javaClass.getResource("/rules/osadl.rules.kts").readText()
+
+            val result = Evaluator(compatibleOrtResult).run(script)
+
+            result.violations should beEmpty()
+        }
+
+        "return a violation for incompatible licenses" {
```

Comment:
Would it make sense to add a test which verifies that a violation is just reported once per dependency / child dependency tuple, e.g. if the dependency appears mutliple times in the dependency tree (maybe also in different scopes)?

Reply:
I basically don't have a problem with adding such a test, but I'm wondering whether this isn't testing too much of an implementation detail of `dependencyRule`:

https://github.com/oss-review-toolkit/ort/blob/22640c2bfe89692bec1a23660ec006469d51f4a0/evaluator/src/main/kotlin/RuleSet.kt#L87-L92

So should the "If the same dependency appears multiple times in the same scope, the rule will be applied only once" bit rather be tested as part of `DependencyRuleTest`?

Reply:
My intention for asking this was to see / point out that this rule flags redundant errors.
In each project, if a dependency with identifier `ID` occurs in `N` scopes I believe the dependency rule is exectued `N` times for `ID`.

Independently of that, I believe the amount of violations for repositories with many definition files, in particular for projects with a fine grained submodule structure can become quite "redundant". So, would it make sense to eliminate the outer "for loop-ish iteration through projects", and replace it by compution the applicable licenses for all (non-excluded) projects, and then using that as outbound license?

Reply:
The `RuleSet` collects all violations in a set, so all duplicates should be eliminated. I think even across projects because it seems that the data in the violations will be identical.

Reply:
> In each project, if a dependency with identifier `ID` occurs in `N` scopes I believe the dependency rule is exectued `N` times for `ID`. What do you think?

I think you're right. IIRC I briefly discussed this with @mnonnenmacher, and we agreed that these "redundant" checks actually make sense, as different scopes might have different distribution channels, and scope excludes might apply.